File: sidecar/content.py
"""Searching the game's own content: which materials and models exist, and what they are.

A mapper picks a texture from Hammer's browser and a prop from its model viewer. An agent
had neither, so it named materials from memory -- and vbsp resolves a name literally, which
makes a wrong one a purple checkerboard nobody sees until a player loads the map. `set_face_
material` can now change a wall's material; this is how you find out what to change it to.

The mounting is the same one `dependencies.py` uses, deliberately: `Game(dir).get_filesystem()`
chains the VPKs and the loose tree in the order the engine reads them. Reusing it means the
answer here and the answer there cannot disagree about what is installed.

Two things this does not do, and says so rather than implying otherwise:

**It does not judge whether a texture looks right.** It reports the shader, the flags and the
size; "is this the brick for a 1970s Bronx tenement" is not a measurement.

**It does not read every VMT to answer a search.** A Garry's Mod install holds tens of
thousands, and parsing them all to answer "brick" would cost seconds per call. The search is
over names; `details` reads the VMTs of what the search returned, which is a handful.
"""

from __future__ import annotations

import fnmatch
import re
from typing import Any

# A search that returns everything is not a search, and serialising 40 000 names through
# the transport would cost more than the caller could use.
DEFAULT_LIMIT = 100
# Animation labels returned per model. A prop has one; a character can have hundreds, and
# a caller asking about a prop should not pay for those.
SEQUENCE_LIMIT = 64
MAX_LIMIT = 1000


def _mount(game_dir: str | None) -> tuple[Any, str | None]:
    if not game_dir:
        return None, "no gameDir given, so nothing was searched"
    try:
        from srctools.game import Game

        return Game(game_dir).get_filesystem(), None
    except Exception as exc:  # noqa: BLE001
        return None, f"{type(exc).__name__}: {exc}"


def _walk(fsys: Any) -> list[str]:
    """Every file the mounted chain can see, as engine-relative paths."""
    out: list[str] = []
    for f in fsys:
        try:
            out.append(f.path.replace("\\", "/"))
        except Exception:  # noqa: BLE001,S112
            continue
    return out


def _searchable(rel: str, prefix: str) -> str:
    """The name a mapper thinks in: `brick/brickwall001a`, no prefix and no extension.

    Matching against the full engine path instead is what a first version did, and it made
    every anchored glob fail silently: `brick/brickwall*` cannot match
    `materials/brick/brickwall001a.vmt`, because fnmatch anchors at both ends and the path
    starts with the prefix. It returned nothing, which is indistinguishable from a game
    that does not have the texture.
    """
    stem = rel[len(prefix) :] if rel.lower().startswith(prefix) else rel
    return stem.rsplit(".", 1)[0].lower()


def _matcher(pattern: str) -> Any:
    """A glob when it looks like one, a substring otherwise.

    `brick*` and `metal/*floor*` are what a mapper types; `brick` on its own means "anything
    with brick in it", not "a file called exactly brick". Guessing wrong in either direction
    returns nothing and looks like the content is missing.

    A glob is tried anchored and then floating, so `brick/brickwall*` matches from the start
    of the name while `*wall00?a` still matches in the middle of it.
    """
    p = pattern.lower().strip("/")
    if any(c in p for c in "*?["):
        return lambda name: fnmatch.fnmatch(name, p) or fnmatch.fnmatch(name, f"*{p}")
    return lambda name: p in name


def _vmf_name(rel: str) -> str:
    """`materials/brick/wall.vmt` as a .vmf stores it: `BRICK/WALL`.

    The form `set_face_material` wants. Returning the path instead would make the caller
    strip the prefix and the suffix itself, which is exactly the step that gets skipped.
    """
    stem = rel[len("materials/") :] if rel.lower().startswith("materials/") else rel
    return re.sub(r"\.vmt$", "", stem, flags=re.IGNORECASE).upper()


def _present(fsys: Any, path: str) -> bool:
    """Whether the mounted chain has this file. `in` is not part of the filesystem API."""
    try:
        fsys[path]
    except (KeyError, FileNotFoundError):
        return False
    except Exception:  # noqa: BLE001
        return False
    return True


def _material_details(fsys: Any, rel: str) -> dict[str, Any]:
    """Shader, base texture, surface properties -- and whether the thing will actually draw.

    That last part is #77, and it is why this function is not just a VMT reader. A `.vmt`
    present in the game says nothing about the `.vtf` it points at. `NATURE/BLENDGRASSDIRT01`
    ships with Garry's Mod; its `forest_grass_01`, `forest_dirt_02` and `blendtexture01` do
    not, and a garden built on it is the purple checkerboard this whole tool exists to
    prevent -- reported as present.

    The resolution is `dependencies.py`'s, imported rather than restated: the two must not be
    able to disagree about what is installed, which is the same reason both mount through
    `Game(dir).get_filesystem()`.
    """
    out: dict[str, Any] = {"shader": None, "basetexture": None, "surfaceprop": None,
                           "translucent": False, "toolTexture": False,
                           "resolves": True, "missingTextures": [], "error": None}
    try:
        from srctools.vmt import Material

        from dependencies import ENVMAP_KEYWORDS, TEXTURE_PARAMS, _under_materials

        with fsys[rel].open_str() as fh:
            mat = Material.parse(fh, rel)
        mat = mat.apply_patches(fsys)
        out["shader"] = mat.shader
        for key in ("$basetexture", "$bumpmap", "$surfaceprop", "$translucent", "$alphatest"):
            val = mat.get(key)
            if val is None:
                continue
            if key == "$basetexture":
                out["basetexture"] = str(val)
            elif key == "$surfaceprop":
                out["surfaceprop"] = str(val)
            elif key in ("$translucent", "$alphatest") and str(val).strip() not in ("0", ""):
                out["translucent"] = True

        # Every texture the material names, not just the base one: a blend draws its second
        # texture wherever alpha is 255, so a missing `$basetexture2` is half a terrain.
        missing: list[str] = []
        seen: set[str] = set()
        for key, value in mat.items():
            low = key.lower()
            if not isinstance(value, str) or not value.strip():
                continue
            if low in TEXTURE_PARAMS or (
                low == "$envmap" and value.lower() not in ENVMAP_KEYWORDS
            ):
                tex = _under_materials(value.strip(), ".vtf")
                if tex in seen:
                    continue
                seen.add(tex)
                if not _present(fsys, tex):
                    missing.append(tex)
        out["missingTextures"] = sorted(missing)
        out["resolves"] = not missing
    except Exception as exc:  # noqa: BLE001
        out["error"] = f"{type(exc).__name__}: {exc}"
        # A VMT this parser cannot read is not evidence that its textures are absent. Saying
        # `resolves: false` there would turn "I could not tell" into "it is broken".
        out["resolves"] = None
    out["toolTexture"] = rel.lower().startswith("materials/tools/")
    return out


def search_content(req: dict[str, Any]) -> dict[str, Any]:
    """Finds materials and models in the game, by name."""
    game_dir = req.get("gameDir")
    pattern = str(req.get("pattern", "")).strip()
    kind = req.get("kind", "material")
    limit = min(int(req.get("limit", DEFAULT_LIMIT)), MAX_LIMIT)
    want_details = bool(req.get("details", False))

    if not pattern:
        # Not a top-level `error`: callSidecar reads any truthy top-level `error` as the
        # fault envelope and looks for `error.message`, so a string there reaches the caller
        # as "unknown error" and this diagnostic is lost. The verb reports the problem in
        # its own result instead.
        return {
            "gameDir": game_dir,
            "mounted": False,
            "mountError": None,
            "pattern": pattern,
            "kind": kind,
            "total": 0,
            "results": [],
            "refused": "a search needs a pattern; an empty one would return the whole game",
        }

    fsys, mount_error = _mount(game_dir)
    if fsys is None:
        return {
            "gameDir": game_dir,
            "mounted": False,
            "mountError": mount_error,
            "pattern": pattern,
            "kind": kind,
            "total": 0,
            "results": [],
            "note": "nothing was searched, so an empty result here says nothing about the game",
        }

    prefix, suffix = (
        ("materials/", ".vmt") if kind == "material" else ("models/", ".mdl")
    )
    match = _matcher(pattern)
    hits: list[str] = []
    for path in _walk(fsys):
        low = path.lower()
        if not low.startswith(prefix) or not low.endswith(suffix):
            continue
        if match(_searchable(low, prefix)):
            hits.append(path)

    hits.sort()
    total = len(hits)
    shown = hits[:limit]

    results: list[dict[str, Any]] = []
    for rel in shown:
        row: dict[str, Any] = {"path": rel}
        if kind == "material":
            row["name"] = _vmf_name(rel)
            if want_details:
                row.update(_material_details(fsys, rel))
        results.append(row)

    return {
        "gameDir": game_dir,
        "mounted": True,
        "mountError": None,
        "pattern": pattern,
        "kind": kind,
        "total": total,
        "shown": len(results),
        "truncated": total > len(results),
        "results": results,
        "note": (
            "names only unless details was asked for: a Garry's Mod install holds tens of "
            "thousands of VMTs and parsing them all to answer one search would cost seconds"
        ),
    }


def model_info(req: dict[str, Any]) -> dict[str, Any]:
    """Bounds, skins, sequences and materials of one .mdl.

    What Hammer's model browser shows, and what an agent needs before placing a prop: a
    prop_static positioned from its origin without knowing its size lands half inside the
    floor, and nothing downstream reports that.
    """
    game_dir = req.get("gameDir")
    rel = str(req.get("model", "")).replace("\\", "/").strip()
    if not rel:
        return {"model": "", "mounted": False, "found": False, "refused": "no model given"}
    if not rel.lower().startswith("models/"):
        rel = f"models/{rel}"
    if not rel.lower().endswith(".mdl"):
        rel = f"{rel}.mdl"

    fsys, mount_error = _mount(game_dir)
    if fsys is None:
        return {"model": rel, "mounted": False, "mountError": mount_error, "found": False}

    try:
        from srctools.mdl import Model

        mdl = Model(fsys, fsys[rel])
    except (KeyError, FileNotFoundError):
        return {
            "model": rel,
            "mounted": True,
            "mountError": None,
            "found": False,
            "note": "not in this game's content. Check the name, or the addon that ships it",
        }
    except Exception as exc:  # noqa: BLE001
        # `readError` rather than `error`, for the same reason as above: a malformed model
        # would otherwise reach the caller as "sidecar model_info: unknown error" with the
        # diagnostic thrown away.
        return {
            "model": rel,
            "mounted": True,
            "mountError": None,
            "found": False,
            "readError": f"{type(exc).__name__}: {exc}",
        }

    def _vec(v: Any) -> list[float] | None:
        try:
            return [round(float(v.x), 3), round(float(v.y), 3), round(float(v.z), 3)]
        except Exception:  # noqa: BLE001
            return None

    textures = []
    try:
        textures = sorted({t.replace("\\", "/") for t in mdl.iter_textures()})
    except Exception:  # noqa: BLE001
        textures = []

    skins: list[int] = []
    try:
        skins = list(range(len(mdl.skins)))
    except Exception:  # noqa: BLE001
        skins = []

    # Labels only. srctools' Sequence reprs its bounding box and its event list, which
    # runs to a few hundred characters each and answers a question nobody asked.
    # Counted before slicing. Reporting the length of the truncated list told a caller a
    # model with 300 animations has exactly 64, with nothing to say the rest exist.
    sequences: list[str] = []
    sequence_total = 0
    try:
        labels = [str(getattr(s, "label", s)) for s in getattr(mdl, "sequences", [])]
        sequence_total = len(labels)
        sequences = labels[:SEQUENCE_LIMIT]
    except Exception:  # noqa: BLE001
        sequences = []

    mins = _vec(getattr(mdl, "hull_min", None))
    maxs = _vec(getattr(mdl, "hull_max", None))
    size = (
        [round(maxs[i] - mins[i], 3) for i in range(3)]
        if mins is not None and maxs is not None
        else None
    )

    return {
        "model": rel,
        "mounted": True,
        "mountError": None,
        "found": True,
        "mins": mins,
        "maxs": maxs,
        "size": size,
        "skinCount": len(skins),
        "sequenceCount": sequence_total,
        "sequences": sequences,
        "sequencesTruncated": sequence_total > len(sequences),
        "materials": textures,
        "note": (
            "bounds are the model's own hull, around its origin. A prop placed at a floor's "
            "height sinks by however far its origin sits above its lowest point"
        ),
    }

File: src/tools/content.ts
import { z } from "zod";
import { gameBlock, gameFor } from "../games/resolve.js";
import { defineTool } from "../mcp/registry.js";
import { callSidecar } from "../sidecar/client.js";
import { GAME, GAME_BLOCK } from "./paths.js";

export const readGameContentTool = defineTool({
  name: "read_game_content",
  description:
    "Searches the game's installed content for materials and models, by name. This is " +
    "Hammer's texture browser and model browser, which an agent otherwise does not have: " +
    "without it a material is named from memory, and vbsp resolves a name literally, so a " +
    "wrong one is a purple checkerboard nobody sees until a player loads the map. Returns " +
    "material names in the form a .vmf stores -- uppercase, no materials/ prefix, no .vmt " +
    "suffix -- which is exactly what set_face_material takes. A pattern with * or ? is a " +
    "glob; anything else is a substring. Mounts the same VPK-and-loose chain the engine " +
    "reads through, so this and read_map_dependencies cannot disagree about what exists. " +
    "A .vmt being present is NOT the same as the material drawing: NATURE/BLENDGRASSDIRT01 " +
    "ships with Garry's Mod and the three .vtf it points at do not, which is the same purple " +
    "checkerboard by another route. Pass details to resolve every texture a material names " +
    "and get `resolves` with what is missing. Needs the Python sidecar; see health.",
  realm: "map",
  inputSchema: {
    pattern: z
      .string()
      .describe("Name to look for. 'brick' matches anything containing it; 'brick*' is a glob."),
    kind: z
      .enum(["material", "model"])
      .default("material")
      .describe("Materials (.vmt) or models (.mdl)."),
    details: z
      .boolean()
      .optional()
      .describe(
        "Read the .vmt of each result for its shader, its surface property, and whether " +
          "every texture it names actually exists in this install. Off by default: a " +
          "Garry's Mod install holds tens of thousands, and parsing them all to answer one " +
          "search would cost seconds.",
      ),
    limit: z.number().int().min(1).max(1000).default(100),
    game: GAME,
  },
  outputSchema: {
    game: GAME_BLOCK,
    gameDir: z.string().nullable(),
    mounted: z.boolean(),
    mountError: z.string().nullable(),
    pattern: z.string(),
    kind: z.string(),
    /** Every match, not only those returned. */
    total: z.number(),
    shown: z.number().optional(),
    truncated: z.boolean().optional(),
    results: z.array(
      z.object({
        path: z.string(),
        /** Materials only: the name as a .vmf writes it. */
        name: z.string().optional(),
        shader: z.string().nullable().optional(),
        basetexture: z.string().nullable().optional(),
        surfaceprop: z.string().nullable().optional(),
        translucent: z.boolean().optional(),
        toolTexture: z.boolean().optional(),
        /**
         * Whether every .vtf this material names is installed. `details` only.
         *
         * `null` means the .vmt could not be parsed, which is not evidence either way --
         * see `error` beside it.
         */
        resolves: z.boolean().nullable().optional(),
        /** The .vtf this material names and the install does not have. `details` only. */
        missingTextures: z.array(z.string()).optional(),
        error: z.string().nullable().optional(),
      }),
    ),
    note: z.string().optional(),
    /** Set when the verb declined the request. Not `error`, which is the fault envelope. */
    refused: z.string().optional(),
  },
  handler: async (args, ctx) => {
    // Refused here rather than in the sidecar so it holds on a machine with no venv: an
    // empty pattern would walk the whole game and serialise tens of thousands of names.
    if (args.pattern.trim().length === 0) {
      throw new Error("a search needs a pattern; an empty one would return the whole game");
    }
    const { game, from } = gameFor(ctx.config, args.game);
    const result = (await callSidecar(
      "search_content",
      {
        // The selected profile's directory, and nothing else. Falling back to Garry's Mod
        // when the asked-for game is not installed returned GMod's assets under a reply
        // that named the other game -- a confident answer about content the map will not
        // have. A null directory makes the sidecar report that nothing was mounted, which
        // is the honest one.
        gameDir: game.gameDir,
        pattern: args.pattern,
        kind: args.kind,
        limit: args.limit,
        ...(args.details !== undefined ? { details: args.details } : {}),
      },
      ctx.config,
      120_000,
    )) as Record<string, unknown>;
    return { ...result, game: gameBlock(game, from) };
  },
});

export const readModelInfoTool = defineTool({
  name: "read_model_info",
  description:
    "Bounds, skins, sequences and materials of one .mdl -- what Hammer's model viewer " +
    "shows before you place a prop. The bounds are the reason this exists: a model's " +
    "origin is wherever the artist put it, so a prop placed at floor height sinks by " +
    "however far its origin sits above its lowest point, and nothing downstream reports " +
    "that. A barrel measured here is 52 units tall with its origin at the centre, which " +
    "means 26 units underground if it is placed on the floor's own z. Needs the Python " +
    "sidecar; see health.",
  realm: "map",
  inputSchema: {
    model: z
      .string()
      .describe("Model path. The models/ prefix and the .mdl suffix are added if missing."),
    game: GAME,
  },
  outputSchema: {
    game: GAME_BLOCK,
    model: z.string(),
    mounted: z.boolean(),
    mountError: z.string().nullable().optional(),
    found: z.boolean(),
    /** Corner of the model's own hull, relative to its origin. */
    mins: z.array(z.number()).nullable().optional(),
    maxs: z.array(z.number()).nullable().optional(),
    size: z.array(z.number()).nullable().optional(),
    skinCount: z.number().optional(),
    /** Every sequence the model has, even when only the first 64 labels are returned. */
    sequenceCount: z.number().optional(),
    sequences: z.array(z.string()).optional(),
    materials: z.array(z.string()).optional(),
    sequencesTruncated: z.boolean().optional(),
    note: z.string().optional(),
    /** A model that exists and could not be read. Not `error`: that is the fault envelope. */
    readError: z.string().optional(),
    refused: z.string().optional(),
  },
  handler: async (args, ctx) => {
    const { game, from } = gameFor(ctx.config, args.game);
    const result = (await callSidecar(
      "model_info",
      // Same rule as read_game_content: the profile's own directory, or nothing.
      { gameDir: game.gameDir, model: args.model },
      ctx.config,
      120_000,
    )) as Record<string, unknown>;
    return { ...result, game: gameBlock(game, from) };
  },
});

export const contentTools = [readGameContentTool, readModelInfoTool];
